arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ColumnVector.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.arrow.vectorized;
+
+import org.apache.arrow.vector.FieldVector;
+import org.apache.iceberg.types.Types;
+
+/**
+ * This class is inspired by Spark's {@code ColumnVector}.
+ * This class represents the column data for an Iceberg table query.
+ * It wraps an arrow {@link FieldVector} and provides simple
+ * accessors for the row values. Advanced users can access
+ * the {@link FieldVector}.
+ * <p>
+ *   Supported Iceberg data types:
+ *   <ul>
+ *     <li>{@link Types.BooleanType}</li>
+ *     <li>{@link Types.IntegerType}</li>
+ *     <li>{@link Types.LongType}</li>
+ *     <li>{@link Types.FloatType}</li>
+ *     <li>{@link Types.DoubleType}</li>
+ *     <li>{@link Types.StringType}</li>
+ *     <li>{@link Types.BinaryType}</li>
+ *     <li>{@link Types.TimestampType} (with and without timezone)</li>
+ *     <li>{@link Types.DateType}</li>
+ *   </ul>
+ */
+public class ColumnVector implements AutoCloseable {
```

Comment:
@rymurr, why did we choose to add a public vector representation? Is Arrow itself insufficient?

Reply:
@mayursrivastava, maybe you could comment on this choice as well?

Reply:
My thinking was to hide the Arrow imp from clients at the cost of an extra layer of indirection. I think the Arrow stuff restricted to the arrow module and out of the public iceberg api will allow more flexibility for future vectorised impls and doesn't couple the two projects too tightly. Happy to discuss if anyone disagrees?

Reply:
My concern is that we're introducing a new API that is specific to Iceberg and will need to be supported for a long time. Not that we can't do that, but I want to make sure we're carefully considering whether we want our own columnar API.

That said, I think there's an argument for doing this because we probably won't be producing Arrow batches from ORC any time soon. @aokolnychyi, what do you think?

Reply:
My use case is to use Arrow VectorSchemaRoot directly, but I agree with @rymurr's suggestions to wrap the arrow data structures. 

I see the following benefits with the wrapper interface:
1. Lifecycle management is better.
2. The current parquet reader returns physical repr of the data as arrow vectors. This means that the dictionary encoded columns are returned as int32 and columns that were widened return the physical file column width (e.g. if int32 was widened to int64 and the data contains int32, the arrow vector is int32). The wrapper classes can handle dictionary encoding and type widening correctly. Note that this is also done in the Spark version. This, however, is a limitation of the current implementation and it will be better for the Arrow Reader to return arrow vectors with logical types.
3. The wrapper interface is easier to use for most users.

The cons are following:
1. A new API is introduced that has to be learned and maintained.

I also don't feel strong about the wrapper interface and if the community doesn't find it useful abstraction, I agree.